calico.go

```diff
@@ -306,57 +306,84 @@ func cmdDel(args *skel.CmdArgs) error {
-		// It just needs a valid CIDR, but it doesn't have to be the CIDR associated with the host.
-		dummyPodCidr := "0.0.0.0/0"
-		var stdinData map[string]interface{}
+	calicoClient, err := CreateClient(conf)
+	if err != nil {
+		return err
+	}
 
-		if err := json.Unmarshal(args.StdinData, &stdinData); err != nil {
-			return err
-		}
+	ep := api.WorkloadEndpointMetadata{
+		Name:             args.IfName,
+		Node:             nodename,
+		Orchestrator:     orchestrator,
+		ActiveInstanceID: "",
+		Workload:         workload,
+	}
 
-		logger.WithField("podCidr", dummyPodCidr).Info("Using a dummy podCidr to release the IP")
-		stdinData["ipam"].(map[string]interface{})["subnet"] = dummyPodCidr
+	var wep *api.WorkloadEndpoint
+	var endpoint api.WorkloadEndpointMetadata
 
-		args.StdinData, err = json.Marshal(stdinData)
+	// The following logic only applies to kubernetes since it sends multiple DELs for the same endpoint.
+	// No need to get WEP from the datastore and compare it's ActiveInstanceID with ContainerID for non-k8s orchestrators
+	if orchestrator == "k8s" {
+		wep, err = calicoClient.WorkloadEndpoints().Get(ep)
 		if err != nil {
-			return err
+			if _, ok := err.(errors.ErrorResourceDoesNotExist); ok {
+
+				// We can talk to the datastore but WEP doesn't exist in there,
+				// but we still want to go ahead with the clean up. So, log a warning and clean up.
+				logger.WithField("WorkloadEndpoint", ep).Warningf("Error getting workloadendpoint: %v", err)
+			} else {
+				// Could not connect to datastore (connection refused, unauthorized, etc.)
+				// so we have no way of knowing/checking ActiveInstanceID. To protect the endpoint
+				// from false DEL, we return the error without deleting/cleaning up.
+				return err
+			}
 		}
-		logger.WithField("stdin", args.StdinData).Debug("Updated stdin data for Delete Cmd")
-	}
 
-	ipamErr := ipam.ExecDel(conf.IPAM.Type, args.StdinData)
+		// Store WEP Metadata in endpoint only if it is not nil to avoid nil pointer dereference
+		// in some of the following operations when WEP doesn't exist in the datastore.
+		if wep != nil {
+			endpoint = wep.Metadata
+		} else {
+			endpoint = ep
+		}
 
-	if ipamErr != nil {
-		logger.Error(ipamErr)
+		// Check if ActiveInstanceID is populated (it will be an empty string "" if it was populated
+		// before this field was added to the API), and if it is there then compare it with ContainerID
+		// passed by the orchestrator to make sure they are the same, return without deleting if they aren't.
+		if endpoint.ActiveInstanceID != "" && args.ContainerID != endpoint.ActiveInstanceID {
+			logger.WithField("WorkloadEndpoint", wep).Warning("CNI_ContainerID does not match WorkloadEndpoint ActiveInstanceID so ignoring the DELETE cmd.")
+			return nil
+		}
 	}
 
-	calicoClient, err := CreateClient(conf)
-	if err != nil {
-		return err
+	// Store WEP Metadata in endpoint only if it is not nil to avoid nil pointer dereference
```

Comment:
yes, it's needed because in case of non-k8s the value of `endpoint` is an empty struct of type `api.WorkloadEndpoint` which won't work, so we need to use the initial `ep` we created to delete the WEP